### ISSUE TITLE
Standardize institution names in responses data frame

### ISSUE THREAD

**Issue by georgeloyer** [georgeloyer/epp#4]
The data input team has adopted standardized institution names in DataInputSheets.  The new standards were adopted after the data input process was started, so there are records in the 'responses' data frame that don't have the standard names.  This may continue to be true for new DataInputSheets. This issue fix should:

1. Update the 'institutionclean' field (there are three variations in 'responses': institutionclean, institutionclean.x and institutionclean.y) with the standard names;
2. Correctly set the institutionclean field where the values are currently NA.
3. Ensure that all of the surveys from a particular class are given the same institutionclean value.
4. Be runnable multiple times without impacting the data set, except to correct any missing or incorrect institutionclean names.
5. Be written so that it can be a component of a data clean up pipeline that runs all of the clean up tasks at once.

The standard names spreadsheet is maintained at: [Naming Conventions](https://docs.google.com/spreadsheets/d/1K6JdHjVMj-1tACrV1bB--GJSuZ-Wq5ygPqIZml9ZXH4/edit#gid=0)

**Comment by georgeloyer** [comment]
The DataCleanup.R file is established with this enhancement. It's used to execute a series of data cleanup R codes that operate on the 'responses' data frame as it comes out of the LoadSurveyData.R import program.  The DataCleanup.R wrapper and all of the subprograms are designed to be idempotent.

The structure of the wrapper is:
`# DataCleanup.R - Main Program

# Define the file paths for the cleanup scripts
cleanup_script1 <- "Cleanup1.R"
cleanup_script2 <- "Cleanup2.R"

# Execute Cleanup1.R
source(cleanup_script1)

# Execute Cleanup2.R
source(cleanup_script2)

# Additional cleanup steps or post-processing code can be added here if needed
`

**Comment by georgeloyer** [comment]
Create the program CleanInstitutionNames.R.  This program requires a sheet file in the "Circle In-Custody Program Outcomes" drive, in the "DATA/1.Data by Course folder named Course_Institutions. The file has two columns that are updated as new courses are input with the standard institution names.  The data quality of the institution names in the google sheet is enforced with a validation drop down that is required.

The program has the following tasks:

- sets up logging to clean_inst logs
- authenticates drive access
- sets up path to Course_Institutions file
- reads data from Course_Institution into a data frame
- opens the 'responses' file and reads into the 'responses' data frame
- extracts the course id embedded in the pkey field and updates the institution_clean field based on the data in Course_Institution data
- reports on course ids that have no Course_Institution data to the console and the log
- if there were any changes made to the responses data frame, write a new responses data file.
- clean up any objects created during the execution of the program

**Comment by georgeloyer** [comment]
This code ends up with a course_id column being added to a modified responses data frame. The LoadSurveyData program is idempotent currently because it tracks what DataInputSheets have been processed and skips them, keeping a record in a processed_files file, and appends the new imported data to the existing responses data frame, which is saved to the responses.Rdata file after a successful import.  Before a successful import run finishes, interim results are saved in temp files so that failures on a single file are handled.

LoadSurveyData depends on responses.Rdata always keeping the same structure, so a cleansing program that changes the structure would break LoadSurveyData. If the cleansing pipeline transforms the responses data frame to a responses_clean data frame with changes in both values and structure, there may be a class of transformations that cannot be reversed.  That will require the cleansing pipeline to track the elements of the pipeline that have been executed on the responses rows so that those elements can be skipped and only newly imported rows be cleansed and then appended to the responses_clean data frame. 

When a transform changes the structure of responses_clean then to be idempotent after responses_clean has been created by other transforms, then the newest transform must detect that it has not yet operated on the existing responses_clean, perform the transform on the existing responses_clean so that it has the new data and the new structure, and then process any new data in responses.

As a safeguard, the original responses data frame must be able to be re-imported from the beginning and all of the transforms applied to create the latest version of responses_clean.  This provides for bugs in the import or the cleansing transforms that are found after the fact and require re-processing to correct the error, rather than in-place data repairs.

The bottom line is that for this issue, the course_id column has value going forward for the analysis and will be kept, while the facility_name will be deleted in favor of the three institution names.  That immediately suggests that a later cleanup will change the columns so that there is no duplication in columns where it isn't bringing some value to the analysis.

This design note should be extracted into the design document for this project.

**Comment by georgeloyer** [comment]
CleanInstitutionNames development has been completed.  Anna has agreed to populate the Courses_Facilities file so I can continue to re-run this program as new data is input.
